### ISSUE TITLE
Support Connecting to localhost servers when served over https

### ISSUE THREAD

**Issue by cycomachead** [jmoenig/Snap#2095]
So this is more discussion than anything....

tl;dr: When you visit https://snap.b.e/ a connection (via the URL block) to http://localhost/my-robot will fail because it doesn't meet security standards.

Problem: It's somewhat involved to setup https for localhost servers. 
Longer Problem: Chrome and others might start blocking requests to localhost

Posting here so we can discuss solutions.

* clients can use self-signed certificates (theoretically, this is a scriptable process)
* Clients can use a "tunnel"ing service like localtunnel.me or or ngrok.io which exposes a local server to a publicly available domain. (This was how Dan and I did the speech "API" at Scratch in Bordeaux.)
* the snap cloud could provide some sort of proxy mechanism to broker requests between a local server and Snap!. This makes implementing a client to be more work.

**Comment by CreativeComputerLab** [comment]
For what its worth, I used 127.0.0.1 instead of localhost and this works fine on most browsers except for Safari.  To expand on your first bullet point I'm contemplating installing my own CA cert into the OS trust store.  That's a lot of work for a little problem.

**Comment by cycomachead** [comment]
Do you mean you can access http://127.0.0.1 from https://snap.berkeley.edu in Chrome?

If so that’s really interesting.

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by cycomachead** [comment]
Even so, I would expect any request to an http domain from an https domain to fail — that’s what’s supposed to happen. If there’s a way around that, I’m not sure if we should count on it, but it might help people in the mean time. 

Also how do people feel about using something like http://robots.snap.berkeley.edu as a page that /only/ serves snap and people can use that as a place to connect to local servers? 
Unfortunately it means that if you open a page on snap.berkeley.edu you would need to save and reload your project. So not ideal, but something doable. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by CreativeComputerLab** [comment]
Yes, in my case I'm using simple http requests with text/plain content. I'm surprised Safari is balking at this.

**Comment by KingDavid12** [comment]
@cycomachead I just tested again now. Chrome and Firefox work fine as long as you stick to plain text and the local server adds CORS headers.

**Comment by cycomachead** [comment]
To be honest, I’m more surprised chrome allows it. 

You aren’t supposed to be able to load any content over http from a page which is loaded with an https connection.  Doing so technically breaks any trust you have in the page because the browser can’t verify what you do with that content. Now for localhost requests, I feel like this a bit on an extreme stance to take but from a security perspective it makes sense. 

The other option is to use browser extensions but this is difficult to build and requires users are able to install software. :/

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by cycomachead** [comment]
Closing because we do have extensions.snap.berkeley.edu
